subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TaskOutputCacheCommandFactory.java

```diff
@@ -103,16 +125,45 @@ public BuildCacheKey getKey() {
             return new BuildCacheLoadCommand.Result<TaskOutputOriginMetadata>() {
                 @Override
                 public long getArtifactEntryCount() {
-                    return unpackResult.entries;
+                    return unpackResult.getEntries();
                 }
 
                 @Override
                 public TaskOutputOriginMetadata getMetadata() {
-                    return unpackResult.originMetadata;
+                    return unpackResult.getOriginMetadata();
                 }
             };
         }
 
+        private void updateSnapshots(ImmutableListMultimap<String, FileSnapshot> propertiesFileSnapshots) {
+            ImmutableMap<String, ResolvedTaskOutputFilePropertySpec> propertySpecsMap = Maps.uniqueIndex(outputProperties, new Function<TaskOutputFilePropertySpec, String>() {
+                @Override
+                public String apply(TaskOutputFilePropertySpec propertySpec) {
+                    return propertySpec.getPropertyName();
+                }
+            });
+            ImmutableSortedMap.Builder<String, FileCollectionSnapshot> propertySnapshotsBuilder = ImmutableSortedMap.naturalOrder();
+            for (String propertyName : propertiesFileSnapshots.keySet()) {
+                ResolvedTaskOutputFilePropertySpec property = propertySpecsMap.get(propertyName);
+                List<FileSnapshot> fileSnapshots = propertiesFileSnapshots.get(propertyName);
+                FileCollectionSnapshotBuilder builder = new FileCollectionSnapshotBuilder(UNORDERED, OutputPathNormalizationStrategy.getInstance(), stringInterner);
+                builder.visitFileTreeSnapshot(fileSnapshots);
```

Comment:
That test is not enough since it tests if the file system mirror has been used correctly. For checking if we recorded the right output snapshot I think we need an up-to-date task as described above. In the current form the test does not fail, but it should since if you run another build for `:test` (which is then up-to-date) right after pulling from the cache, an overlapping output is detected (the output directory), which means that we did not record the right output snapshot.